File: Editor/MapEditor/floor.cpp
/*
    RPG Paper Maker Copyright (C) 2017-2022 Wano

    RPG Paper Maker engine is under proprietary license.
    This source code is also copyrighted.

    Use Commercial edition for commercial use of your games.
    See RPG Paper Maker EULA here:
        http://rpg-paper-maker.com/index.php/eula.
*/

#include "floor.h"
#include "lands.h"
#include "rpm.h"

// -------------------------------------------------------
//
//
//  ---------- FLOORDATAS
//
//
// -------------------------------------------------------

// -------------------------------------------------------
//
//  CONSTRUCTOR / DESTRUCTOR / GET / SET
//
// -------------------------------------------------------

FloorDatas::FloorDatas() :
    FloorDatas(new QRect)
{

}

FloorDatas::FloorDatas(QRect* texture, bool up) :
    LandDatas(texture, up)
{

}

FloorDatas::~FloorDatas()
{

}

bool FloorDatas::operator==(const FloorDatas& other) const {
    return LandDatas::operator==(other);
}

bool FloorDatas::operator!=(const FloorDatas& other) const {
    return !operator==(other);
}

MapEditorSubSelectionKind FloorDatas::getSubKind() const{
    return MapEditorSubSelectionKind::Floors;
}

QString FloorDatas::toString() const {
    return RPM::translate(Translations::FLOOR).toUpper();
}

// -------------------------------------------------------
//
//  INTERMEDIARY FUNCTIONS
//
// -------------------------------------------------------

void FloorDatas::initializeVertices(int squareSize, int width, int height,
                                   QVector<Vertex>& vertices,
                                   QVector<GLuint>& indexes, Position& position,
                                   int& count)
{
    QVector3D pos, size, center;
    getPosSizeCenterLand(pos, size, center, squareSize, position);

    float x = (float)(textureRect()->x() * squareSize) / width;
    float y = (float)(textureRect()->y() * squareSize) / height;
    float w = (float)(textureRect()->width() * squareSize) / width;
    float h = (float)(textureRect()->height() * squareSize) / height;
    if (w <= 0 || h <= 0)
    {
        return;
    }
    float coefX = 0.1 / width;
    float coefY = 0.1 / height;
    x += coefX;
    y += coefY;
    w -= (coefX * 2);
    h -= (coefY * 2);

    // Vertices
    QVector3D vecA = Lands::verticesQuad[0] * size + pos;
    QVector3D vecB = Lands::verticesQuad[1] * size + pos;
    QVector3D vecC = Lands::verticesQuad[2] * size + pos;
    QVector3D vecD = Lands::verticesQuad[3] * size + pos;
    if (position.angleY() != 0.0) {
        SpriteDatas::rotateSpriteX(vecA, vecB, vecC, vecD, center, position
            .angleY(), 0, 1, 0);
    }
    if (position.angleX() != 0.0) {
        SpriteDatas::rotateSpriteX(vecA, vecB, vecC, vecD, center, position
            .angleX(), 1, 0, 0);
    }
    if (position.angleZ() != 0.0) {
        SpriteDatas::rotateSpriteX(vecA, vecB, vecC, vecD, center, position
            .angleZ(), 0, 0, 1);
    }
    vertices.append(Vertex(vecA, QVector2D(x, y)));
    vertices.append(Vertex(vecB, QVector2D(x + w, y)));
    vertices.append(Vertex(vecC, QVector2D(x + w, y + h)));
    vertices.append(Vertex(vecD, QVector2D(x, y + h)));

    // indexes
    int offset = count * Lands::nbVerticesQuad;
    for (int i = 0; i < Lands::nbIndexesQuad; i++)
        indexes.append(Lands::indexesQuad[i] + offset);

    count++;
}

// -------------------------------------------------------
//
//  READ / WRITE
//
// -------------------------------------------------------

void FloorDatas::read(const QJsonObject & json){
    LandDatas::read(json);
}

// -------------------------------------------------------

void FloorDatas::write(QJsonObject &json) const{
    LandDatas::write(json);
}

File: src/editor/mapElements/Floor.ts
/*
    RPG Paper Maker Copyright (C) 2017-2026 Wano

    RPG Paper Maker engine is under proprietary license.
    This source code is also copyrighted.

    Use Commercial edition for commercial use of your games.
    See RPG Paper Maker EULA here:
        http://rpg-paper-maker.com/index.php/eula.
*/

import i18next from 'i18next';
import { ELEMENT_MAP_KIND, JSONType, Mathf } from '../common';
import { CustomGeometry } from '../core/CustomGeometry';
import { Position } from '../core/Position';
import { Project } from '../core/Project';
import { Rectangle } from '../core/Rectangle';
import { MapElement, Scene } from '../Editor';
import { Land } from './Land';

const { t } = i18next;

class Floor extends Land {
	constructor() {
		super();
		this.kind = ELEMENT_MAP_KIND.FLOOR;
	}

	static fromJSON(json: JSONType): Floor {
		const floor = new Floor();
		floor.read(json);
		return floor;
	}

	static create(texture: Rectangle, up = true): Floor {
		const floor = new Floor();
		floor.texture = texture;
		floor.up = up;
		return floor;
	}

	static getTextureReduced(texture: Rectangle, x: number, z: number): Rectangle {
		const rect = new Rectangle();
		rect.x = texture.x + Mathf.mod(x, texture.width);
		rect.y = texture.y + Mathf.mod(z, texture.height);
		return rect;
	}

	equals(mapElement: MapElement.Base) {
		return super.equals(mapElement);
	}

	toString(): string {
		return `${t('floor')} ${this.texture.toString(true)}`;
	}

	updateGeometry(
		map: Scene.Map,
		geometry: CustomGeometry,
		position: Position,
		width: number,
		height: number,
		count: number,
		forceOffset = false,
	) {
		if (width !== 0 && height !== 0) {
			super.updateGeometryLand(
				map,
				geometry,
				position,
				width,
				height,
				(this.texture.x * Project.SQUARE_SIZE) / width,
				(this.texture.y * Project.SQUARE_SIZE) / height,
				(this.texture.width * Project.SQUARE_SIZE) / width,
				(this.texture.height * Project.SQUARE_SIZE) / height,
				count,
				this.texture.width,
				this.texture.height,
				forceOffset,
			);
		}
	}
}

export { Floor };
